src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,169 @@
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($bucket);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        $this->bucket->ensureIndexes();
+
+        if (!is_resource($source) || get_resource_type($source) != "stream") {
+            throw new UnexpectedTypeException('stream', $source);
+        } else{
+            $streamMetadata = stream_get_meta_data($source);
+        } if (!is_readable($streamMetadata['uri'])) {
+     //       throw new InvalidArgumentException("stream not readable");
+            //issue being that php's is_readable reports native streams as not readable like php://temp
+        }
+        while ($data = fread($source, $this->bucket->getChunkSizeBytes())) {
+            $this->insertChunk($data);
+        }
+        return $this->fileCollectionInsert();
+    }
+    /**
+    * Insert a chunks into GridFS from a string
+    *
+    * @param string  $toWrite   String to upload
+    * @return int
+    */
+    public function insertChunks($toWrite)
+    {
+        $this->bucket->ensureIndexes();
+
+        $readBytes = 0;
+        while($readBytes != strlen($toWrite)) {
+            $addToBuffer = substr($toWrite, $readBytes, $this->bucket->getChunkSizeBytes() - $this->bufferLength);
+            fwrite($this->buffer, $addToBuffer);
+            $readBytes += strlen($addToBuffer);
+            $this->bufferLength += strlen($addToBuffer);
+            if($this->bufferLength == $this->bucket->getChunkSizeBytes()) {
+                rewind($this->buffer);
+                $this->insertChunk(fread($this->buffer, $this->bucket->getChunkSizeBytes()));
+                ftruncate($this->buffer,0);
+                $this->bufferLength = 0;
+            }
+        }
+        return $readBytes;
+    }
+    /**
+    * Close an active stream, pushes all buffered data to GridFS
+    *
+    */
+    public function close()
+    {
+        rewind($this->buffer);
+        $cached = fread($this->buffer, $this->bucket->getChunkSizeBytes());
+
+        if(strlen($cached) > 0) {
+            $toUpload = ["files_id" => $this->file["_id"], "n" => $this->n, "data" => $cached];
+            $this->bucket->chunkInsert($toUpload);
+            $this->n++;
+        }
+
+        fclose($this->buffer);
+
+        $this->fileCollectionInsert();
+    }
+    private function insertChunk($data)
+    {
+        $toUpload = ["files_id" => $this->file['_id'], "n" => $this->n, "data" => $data];
+        hash_update($this->ctx, $data);
+        $this->bucket->chunkInsert($toUpload);
```

Comment:
Is `Bucket::chunkInsert()` called from anywhere other than GridFsUpload? It seems like that logic could just exist within this class when we move to injecting the Manager and using `executeBulkWrite()` directly.